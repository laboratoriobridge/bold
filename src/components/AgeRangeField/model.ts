export enum AgeRangeUnitEnum {
  YEARS,
  MONTHS,
  DAYS,
}

export interface AgeRange {
  firstValue?: number
  secondValue?: number
  unit: AgeRangeUnitEnum
}

export const UNIT_OPTIONS = [AgeRangeUnitEnum.YEARS, AgeRangeUnitEnum.MONTHS, AgeRangeUnitEnum.DAYS]
