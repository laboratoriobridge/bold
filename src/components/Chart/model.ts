import { Color } from 'csstype'
import { CSSProperties } from 'react'
import { Moment } from 'moment'
import { blue, gray, orange, pink } from '../../styles/colors'

const CHART_COLOR_SCHEMES = {
  blue: Object.values(blue).filter((_, i) => i % 2) as Color[],
  orange: Object.values(orange).filter((_, i) => i % 2) as Color[],
  pink: Object.values(pink).filter((_, i) => i % 2) as Color[],
  gray: Object.values(gray).filter((_, i) => i % 2) as Color[],
  default: [blue.c40, orange.c50, blue.c50, pink.c40, orange.c60, blue.c20, pink.c50],
}

export type ChartColorScheme = keyof typeof CHART_COLOR_SCHEMES | Color[]
export type ValueRange = { init: number; end: number; step?: number }
export type UnitOfTime = 'year' | 'month' | 'week' | 'day' | 'hour' | 'minute' | 'second' | 'millisecond'
export type DateRangeStep = { amount: number; unit: UnitOfTime }
export type DateRange = {
  init: Date
  end: Date
  format?: (date: Date) => string
  step?: DateRangeStep
}
export type SeriesDataPoint = number | Date | DataPoint<number> | DataPoint<Date>
export type ReferenceAreaDataPoint = number | DataPoint<number>
export type AxisDomain = string[] | ValueRange | DateRange
export type TooltipType = 'point' | 'line' | 'none'
export type DotShape = 'circle' | 'square' | 'rect' | 'triangle' | 'diamond' | 'cross' | 'star' | 'happy'

export type TooltipRenderer<XDomain> = (
  points?: (DataPoint<XDomain> & { seriesName: string })[]
) => string | React.ReactNode

export enum SeriesType {
  Line,
  Column,
  Area,
}

export interface DataPoint<XDomain, YDomain = number> {
  x: XDomain
  y: YDomain
}

export type ChartSeriesDataPoint<XDomain> = number | DataPoint<XDomain, number>

export interface ChartSeries<XDomain> {
  type?: SeriesType
  name: string
  data: ChartSeriesDataPoint<XDomain>[] | boolean[]
  color?: string
  dashed?: boolean
  dot?: false | DotShape
  dataKey?: string
}

export interface BarChartSeries<YDomain> {
  name: string
  data: number[] | DataPoint<number, YDomain>[]
  color?: string
}

export interface PieChartDataPoint {
  name: string
  value: number
  color?: string
}

export interface RangeArea<XDomain> {
  name: string
  init: XDomain
  end: XDomain
  strokeColor?: string | false
  fillColor?: string | false
  fillOpacity?: number
  tickColor?: string | false
}

export interface ReferenceAreaRange<XDomain> {
  x: XDomain
  upperLimit?: number | 'yEnd' | 'yInit'
}

export interface ReferenceAreaPercent<XDomain> {
  x: XDomain
  percent: number
}

interface ReferenceAreaDescription {
  text: string
  color?: string
  style?: CSSProperties
  align?: 'top' | 'bottom'
}

export interface ReferenceArea<XDomain> {
  name: string
  description?: ReferenceAreaDescription
  textYOffset?: number
  area: ReferenceAreaRange<XDomain>[]
  color?: string
  tickColor?: string
  stroke?: boolean
  strokeColor?: string
  strokeType?: 'line' | 'dashed'
}

export interface ReferenceAreaWithPercents<XDomain> extends ReferenceArea<XDomain> {
  areaPercents: ReferenceAreaPercent<XDomain>[]
}

export interface AxisOptions {
  title?: string
  unit?: string
  domain?: AxisDomain
  tickRenderer?: (
    tick: TickProps,
    domainMaxValue: number | Moment | Date,
    isOutlierIndicator?: boolean
  ) => React.SVGProps<SVGElement>
}

export interface RangeSelectorOptions {
  label?: string
  options: { [x: string]: Partial<AxisDomain> }
  defaultOption?: string
}

export interface TickPayload {
  coordinate: number
  isShow: boolean
  offset: number
  tickCoord: number
  value: any
}

export interface TickProps {
  x?: number
  y?: number
  height?: number
  payload?: TickPayload
  fill?: string
  stroke?: string
  textAnchor?: string
  width?: number
  isOutlierIndicator?: boolean
  domainMaxValue?: number | Date
}

export interface CustomDotProps {
  cx?: number
  cy?: number
  stroke?: string
  shape?: DotShape
}

export interface TooltipOptions<XDomain> {
  type: TooltipType
  render?: TooltipRenderer<XDomain>
}

export interface DataPointWithOutlier<XDomain> {
  data: ChartSeriesDataPoint<XDomain>
  isOutlier: boolean
}

export function isValueRange(x: AxisDomain): x is ValueRange {
  return typeof (x as ValueRange).init === 'number'
}

export function isDateRange(x: AxisDomain): x is DateRange {
  return Object.prototype.toString.call((x as DateRange).init) === '[object Date]'
}

export function getChartColorScheme(colorScheme: ChartColorScheme): Color[] {
  if (Array.isArray(colorScheme)) return colorScheme
  return CHART_COLOR_SCHEMES[colorScheme]
}
