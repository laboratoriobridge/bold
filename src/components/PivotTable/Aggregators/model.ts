import { LocaleConfiguration } from '../../../i18n/LocaleContext'

export enum AggregatorEnum {
  PERCENTAGE = 'PERCENTAGE',
  COUNT = 'COUNT',
  AVERAGE = 'AVERAGE',
  MAXIMUM = 'MAXIMUM',
  MINIMUM = 'MINIMUM',
}

export type AggregatorLabels = Pick<
  LocaleConfiguration['aggregators'],
  'count' | 'percentage' | 'average' | 'maximum' | 'minimum'
>

export type AggregatorFunction = (values: number[], total?: number) => number

export type Aggregator = {
  id: AggregatorEnum
  label: string
  value: AggregatorFunction | undefined
  keyDependent: boolean
  chain?: Aggregator[]
  suffix?: string
}
