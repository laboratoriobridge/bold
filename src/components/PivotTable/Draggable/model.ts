import { KeyNavigationDirection } from '../Droppable/types/model'

export const keyDirectionMap: Record<string, KeyNavigationDirection> = {
  ArrowRight: 'right',
  ArrowLeft: 'left',
  ArrowUp: 'up',
  ArrowDown: 'down',
}
