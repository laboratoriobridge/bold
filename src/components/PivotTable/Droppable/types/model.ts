export type KeyNavigationDirection = 'left' | 'right' | 'up' | 'down'
