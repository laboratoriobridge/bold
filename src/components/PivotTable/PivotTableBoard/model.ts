import { LocaleConfiguration } from '../../../i18n'

export type PivotTableBoardLabels = LocaleConfiguration['pivotTableBoard']

export type FieldValuesByKey<T> = Map<keyof T, Array<string>>

export type FieldFiltersByKey<T> = Map<keyof T, Set<string>>

export type RowColumnKeys<T> = [Array<keyof T>, Array<keyof T>]

export type PivotTableBoardOrigin = 'filter' | 'row' | 'column'

export type BoardField<T> = {
  key: keyof T
  origin: PivotTableBoardOrigin
  filters: string[]
}
