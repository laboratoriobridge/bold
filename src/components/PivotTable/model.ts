export type KeyConfig = {
  keyName: string
  formatter?: (value: string) => string
  ordenator?: (a: string, b: string) => number
}

export type KeyMap<T extends object> = Map<keyof T, KeyConfig>
