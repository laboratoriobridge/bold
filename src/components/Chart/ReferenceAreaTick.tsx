import React from 'react'
import { Rectangle } from 'recharts'

import { useTheme } from '../../styles'
import { splitIntoLines } from '../../util/string'
import { ReferenceAreaWithPercents, TickProps } from './model'

const MAX_CHARS_PER_LINE = 8
const TICK_MARGIN = 5
const TICK_X_DISLOCATION = 15
const TICK_Y_DISLOCATION = 15

export interface ReferenceTickProps extends TickProps {
  refTicks: Map<number, ReferenceAreaWithPercents<any>>
}

export function ReferenceAreaTick(props: ReferenceTickProps) {
  const { x, y, payload, refTicks, height } = props

  const ref = refTicks.get(+payload.value)

  const theme = useTheme()

  const nameLines = splitIntoLines(ref.name, MAX_CHARS_PER_LINE)
  const description = ref.description

  return (
    <>
      <text
        x={x}
        y={y}
        dx={TICK_X_DISLOCATION}
        dy={TICK_Y_DISLOCATION}
        textAnchor='start'
        fill={ref.tickColor ?? ref.color}
        style={{ fontWeight: 'bold' }}
      >
        {nameLines.map(
          (namePart, i) =>
            namePart && (
              <tspan key={namePart} dx={TICK_X_DISLOCATION} dy={(i + 1) * TICK_Y_DISLOCATION} x={x} y={y}>
                {namePart}
              </tspan>
            )
        )}
      </text>
      {description && (
        <text
          x={x + TICK_X_DISLOCATION}
          y={y + 5 + nameLines.length * TICK_Y_DISLOCATION}
          dx={TICK_X_DISLOCATION}
          dy={TICK_Y_DISLOCATION}
          textAnchor='start'
          fill={description.color ?? theme.pallete.text.main}
          style={description.style}
        >
          {splitIntoLines(description.text, MAX_CHARS_PER_LINE).map(
            (descriptionPart, i) =>
              descriptionPart && (
                <tspan
                  key={descriptionPart}
                  dx={TICK_X_DISLOCATION}
                  dy={(i + 1) * TICK_Y_DISLOCATION}
                  x={x}
                  y={y + 5 + nameLines.length * TICK_Y_DISLOCATION}
                >
                  {descriptionPart}
                </tspan>
              )
          )}
        </text>
      )}
      <Rectangle
        x={x}
        y={y}
        dx={15}
        dy={TICK_MARGIN / 2}
        width={4}
        height={(ref.areaPercents.slice(-1)[0].percent / 100) * height - TICK_MARGIN / 2}
        fill={ref.tickColor ?? ref.color}
      />
    </>
  )
}
