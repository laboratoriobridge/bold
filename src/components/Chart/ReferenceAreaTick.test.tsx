import React from 'react'
import { render } from '@testing-library/react'
import { colors } from '../../styles'
import { ReferenceAreaTick, ReferenceAreaTickProps } from './ReferenceAreaTick'
import { ReferenceAreaWithPercents } from './model'

const defaultReferenceAreaTickProps: Partial<ReferenceAreaTickProps> = {
  x: 10,
  y: 20,
  payload: { value: 1, coordinate: 0, isShow: true, offset: 0, tickCoord: 0 },
  height: 100,
}

const defaultReferenceAreaWithPercents: ReferenceAreaWithPercents<string> = {
  label: { name: 'title' },
  area: [],
  color: colors.blue.c10,
  tick: {},
  stroke: {},
  areaPercents: [{ x: 'x', percent: 10 }],
}

describe('ReferenceAreaTick', () => {
  it('should renders correctly with the default props', () => {
    const { container } = render(
      <ReferenceAreaTick
        refTicks={new Map([[1, defaultReferenceAreaWithPercents]])}
        {...defaultReferenceAreaTickProps}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should renders correctly when label alignment is central', () => {
    const { container } = render(
      <ReferenceAreaTick
        refTicks={
          new Map([
            [
              1,
              {
                ...defaultReferenceAreaWithPercents,
                label: { name: 'name', description: 'description', alignment: 'central' },
              },
            ],
          ])
        }
        {...defaultReferenceAreaTickProps}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should renders correctly when label alignment is central and name or description has more than one line', () => {
    const { container } = render(
      <ReferenceAreaTick
        refTicks={
          new Map([
            [
              1,
              {
                ...defaultReferenceAreaWithPercents,
                label: { name: 'firstLine secondLine', description: 'fisrtLine secondLine', alignment: 'central' },
              },
            ],
          ])
        }
        {...defaultReferenceAreaTickProps}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should renders correctly when set label color', () => {
    const { container } = render(
      <ReferenceAreaTick
        refTicks={
          new Map([[1, { ...defaultReferenceAreaWithPercents, label: { name: 'name', color: colors.red.c10 } }]])
        }
        {...defaultReferenceAreaTickProps}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should renders correctly when tick kind is horizontal', () => {
    const { container } = render(
      <ReferenceAreaTick
        refTicks={new Map([[1, { ...defaultReferenceAreaWithPercents, tick: { kind: 'horizontal' } }]])}
        {...defaultReferenceAreaTickProps}
      />
    )

    expect(container).toMatchSnapshot()
  })
})
