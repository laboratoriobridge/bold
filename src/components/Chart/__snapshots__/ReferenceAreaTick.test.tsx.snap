// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferenceAreaTick should renders correctly when label alignment is central 1`] = `
<div>
  <text
    dominant-baseline="central"
    dx="15"
    dy="0"
    fill="#002451"
    style="font-weight: bold;"
    text-anchor="start"
    x="10"
    y="20"
  >
    <tspan
      dx="15"
      dy="0"
      x="10"
      y="20"
    >
      name
    </tspan>
  </text>
  <text
    dominant-baseline="central"
    dx="15"
    dy="0"
    fill="#24252E"
    text-anchor="start"
    x="25"
    y="40"
  >
    <tspan
      dx="15"
      dy="0"
      x="10"
      y="40"
    >
      description
    </tspan>
  </text>
  <path
    class="recharts-rectangle"
    d="M 10,20 h 4 v 7.5 h -4 Z"
    dx="15"
    dy="2.5"
    fill="#002451"
    height="7.5"
    radius="0"
    width="4"
    x="10"
    y="20"
  />
</div>
`;

exports[`ReferenceAreaTick should renders correctly when label alignment is central and name or description has more than one line 1`] = `
<div>
  <text
    dominant-baseline="central"
    dx="15"
    dy="0"
    fill="#002451"
    style="font-weight: bold;"
    text-anchor="start"
    x="10"
    y="12.5"
  >
    <tspan
      dx="15"
      dy="0"
      x="10"
      y="12.5"
    >
      firstLine
    </tspan>
    <tspan
      dx="15"
      dy="15"
      x="10"
      y="12.5"
    >
      secondLine
    </tspan>
  </text>
  <text
    dominant-baseline="central"
    dx="15"
    dy="0"
    fill="#24252E"
    text-anchor="start"
    x="25"
    y="47.5"
  >
    <tspan
      dx="15"
      dy="0"
      x="10"
      y="47.5"
    >
      fisrtLine
    </tspan>
    <tspan
      dx="15"
      dy="15"
      x="10"
      y="47.5"
    >
      secondLine
    </tspan>
  </text>
  <path
    class="recharts-rectangle"
    d="M 10,20 h 4 v 7.5 h -4 Z"
    dx="15"
    dy="2.5"
    fill="#002451"
    height="7.5"
    radius="0"
    width="4"
    x="10"
    y="20"
  />
</div>
`;

exports[`ReferenceAreaTick should renders correctly when set label color 1`] = `
<div>
  <text
    dominant-baseline="text-before-edge"
    dx="15"
    dy="0"
    fill="#510609"
    style="font-weight: bold;"
    text-anchor="start"
    x="10"
    y="20"
  >
    <tspan
      dx="15"
      dy="0"
      x="10"
      y="20"
    >
      name
    </tspan>
  </text>
  <path
    class="recharts-rectangle"
    d="M 10,20 h 4 v 7.5 h -4 Z"
    dx="15"
    dy="2.5"
    fill="#002451"
    height="7.5"
    radius="0"
    width="4"
    x="10"
    y="20"
  />
</div>
`;

exports[`ReferenceAreaTick should renders correctly when tick kind is horizontal 1`] = `
<div>
  <text
    dominant-baseline="text-before-edge"
    dx="15"
    dy="0"
    fill="#002451"
    style="font-weight: bold;"
    text-anchor="start"
    x="10"
    y="20"
  >
    <tspan
      dx="15"
      dy="0"
      x="10"
      y="20"
    >
      title
    </tspan>
  </text>
  <path
    class="recharts-rectangle"
    d="M 10,20 h 8 v 2 h -8 Z"
    dx="15"
    dy="2.5"
    fill="#002451"
    height="2"
    radius="0"
    width="8"
    x="10"
    y="20"
  />
</div>
`;

exports[`ReferenceAreaTick should renders correctly with the default props 1`] = `
<div>
  <text
    dominant-baseline="text-before-edge"
    dx="15"
    dy="0"
    fill="#002451"
    style="font-weight: bold;"
    text-anchor="start"
    x="10"
    y="20"
  >
    <tspan
      dx="15"
      dy="0"
      x="10"
      y="20"
    >
      title
    </tspan>
  </text>
  <path
    class="recharts-rectangle"
    d="M 10,20 h 4 v 7.5 h -4 Z"
    dx="15"
    dy="2.5"
    fill="#002451"
    height="7.5"
    radius="0"
    width="4"
    x="10"
    y="20"
  />
</div>
`;
